src/Common/src/System/Collections/Generic/EnumerableHelpers.cs

```diff
@@ -2,23 +2,220 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 // See the LICENSE file in the project root for more information.
 
+using System.Diagnostics;
+
 namespace System.Collections.Generic
 {
     /// <summary>Internal helper functions for working with enumerables.</summary>
     internal static class EnumerableHelpers
     {
-        /// <summary>Converts an enumerable to an array using the same logic as does List{T}.</summary>
+        /// <summary>Converts an enumerable to an array.</summary>
         /// <param name="source">The enumerable to convert.</param>
         /// <returns>The resulting array.</returns>
         internal static T[] ToArray<T>(IEnumerable<T> source)
         {
-            int count;
-            T[] results = ToArray(source, out count);
-            Array.Resize(ref results, count);
-            return results;
+            Debug.Assert(source != null);
+
+            var collection = source as ICollection<T>;
+            if (collection != null)
+            {
+                int count = collection.Count;
+                if (count == 0)
+                    return Array.Empty<T>();
+                
+                var result = new T[count];
+                collection.CopyTo(result, arrayIndex: 0);
+                return result;
+            }
+
+            // Generic methods are compiled per-instantiation for value types.
+            // Do not force the jit to compile a bunch of code it will never
```

Comment:
> This trick does not work for AOT - you will pay for the extra code there.

How much is the cost? Since the method is compiled in advance, there is no first-time overhead invoking the method like when it is jitted.